resource: add support for node-local storage to hwloc reader

SteVwonder opened the pull request:
Problem: versions 2+ of hwloc will emit OS device objects that represent
node-local block devices, but the resource hwloc reader does not pick
those up when they exist.

Solution: add a new check for `HWLOC_OBJ_OSDEV_BLOCK` in the
`walk_hwloc` method and parse out the size from the hwloc object
properties.  The name of the resource in the graph is `storage`.

Also adds tests using hwloc output from three LC clusters that
have node-local storage.

SteVwonder commented:
Looks like the tests are failing on hwloc v1 images.  Unsurprising since I only tested on focal.  I'll dig down into the errors tomorrow.

SteVwonder commented:
Travis is green 🥳 (Github just doesn't know it yet).  I think this PR is ready for review now.

dongahn commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -247,8 +249,21 @@ void resource_reader_hwloc_t::walk_hwloc (resource_graph_t &g,
-            supported_resource = false;
+            supported_resource = true;
+        } else if (obj->attr->osdev.type == HWLOC_OBJ_OSDEV_BLOCK) {
+            if (!in_allowlist ("storage")) {
```
In terms of storage resource vertex names, will `storage[0]->chunk[k]` work with where job spec is headed where `storage` is mapped to each block device and each `chunk` within it maps to a scheduleable unit pool? Or would flat `storage_chunk[l]` make more sense? I like the hierarchical specification as our pruning can take advantage of this for performance. 

How about memory? we can perhaps use `memory[0]->chunk[m]` similarly?

This also seems to map well to other multi-tiered storage...

SteVwonder replied:
> What is granularity at which hwloc reports a block device? Is it for entire device?

Based on the hwloc source code, it is poking at `/sys/class/block` to see which block devices are on the node.  So whatever linux wants to report as a block device, and the granulatiry that linux uses, is what is going to propagate up through hwloc into Flux.  In the case of Lassen and Corona, this means just a single device/resource.  In the case of Catalyst, there are multiples devices (that are presented as a single mounted filesystem....maybe they are ZFS?).

> I think it would be a bit better to get the granularity from a config file and create logical scheduleable chunk.

I agree. Having that flexibility is going to be key for workflows that want multiple jobs running on a node where each job only uses a piece of the storage.  Note that AFAIK, there are no mechanisms in cgroups to enforce this capacity restrict.  I think we will need to do partitioning if we want a "hard enforcement" rather than just the "soft check" of scheduling.

>  will storage[0]->chunk[k] work with where job spec is headed where storage is mapped to each block device and each chunk within it maps to a scheduleable unit pool?
> How about memory? we can perhaps use memory[0]->chunk[m] similarly?

I think those two suggestions make sense given the current feature set of Fluxion.  I think it will be cool if we can eventually get to a point where you can get partial exclusive allocations on a resource (e.g., 512 GB of the 1500 GB on a drive).  That is much further down the road though.

SteVwonder replied:
> although maybe a descriptor like unit would make it even more (perhaps too) generic.

That would be where I ultimately want to go too.  The issue is that right now in Fluxion, only one job can have exclusive access to a resource at a time.  There is no way to ask for exclusive access to a fraction of the capacity of a resource AND still have other jobs run on that resource.  We will need to do some foundational work in the resource model/traversers to make that work.  @dongahn can correct me if I'm off base with my understanding.

SteVwonder replied:
Opened an issue to track this suggestion: https://github.com/flux-framework/flux-sched/issues/722

dongahn replied:
> The issue is that right now in Fluxion, only one job can have exclusive access to a resource at a time.

You can have shared access like `rank[2]->node[4]` where your job has shared access to 2 racks in a shared fashion. What is not there is to have multiple "exclusive accesses" on the resource vertex's own quantity.

To support the latter, you will need to put a bit more metadata etc. Haven't thought much about this because I thought we can always refined the granularity of such pool resource by introducing more vertices.

dongahn commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -247,8 +249,21 @@ void resource_reader_hwloc_t::walk_hwloc (resource_graph_t &g,
+            // Size from hwloc is in kBs (base 10)
+            const char* size_str = hwloc_obj_get_info_by_name (obj, "Size");
+            long int num_bytes = strtol (size_str, NULL, 10);
+            if (size_str == NULL || num_bytes == 0) {
```
error path:

- `strtol` can return other return values (`LONG_MIN` and `LONG_MAX`) to indicate errors. We need to cover this.

- Does `hwloc_obj_get_info_by_name` set errno? Otherwise we may need to set `errno=ENOENT` at this level.

`walk_hwloc` currently doesn't return an error (`void` return type). We many want to change the return type to `int` to pass the error condition to upper frames. 

If the conditional is non-recoverable, this function should return -1 with `errno` set appropriately.

On -1, the parent frame should either transparently return the same rc (or add -1 to rc if it also has an error). This way, the upper layer (e.g., fluxion resource module) can detect this condition and fail the initialization? Or try to recover.

dongahn commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -247,8 +249,21 @@ void resource_reader_hwloc_t::walk_hwloc (resource_graph_t &g,
+            }
+            size = num_bytes / 1000000; // kBs -> gBs
+            type = "storage";
+            basename = type;
```
Should this be set to the device name? Is it available from hwloc?

SteVwonder replied:
The only options available from hwloc are the extended attributes `Vendor`, `Model`, `SerialNumber`, but these aren't always nice looking or present:

Catalyst:
```xml
                <info name="Vendor" value="INTEL_R_"/>
                <info name="Model" value="SSD_910_200GB"/>
                <info name="SerialNumber" value="5000cca013017944"/>
```

Corona:
```xml
                  <info name="Model" value="HUSMR7616BDP301"/>
                  <info name="SerialNumber" value="SDM00001BE7D"/>
```

Lassen:
```xml
                <info name="Model" value="LVM PV TdlTim-H0IX-TkeP-WtpY-Gine-YoJd-ChlfL9 on /dev/nvme0n1"/>
                <info name="SerialNumber" value="S3RVNA0K301124"/>
```

There is also the `PCIVendor` and `PCIDevice` that we can try and pull from.  They appear to be much nicer looking, but there is not a guarantee that node-local storage will be via PCI (although its more likely than not). There is also no guarantee whether the drive vendor appears in the `PCIVendor` space or the Drive's `Vendor` space:

Catalyst:
```xml
              <info name="PCIVendor" value="Broadcom / LSI"/>
              <info name="PCIDevice" value="SAS2008 PCI-Express Fusion-MPT SAS-2 [Falcon]"/>
```

Corona:
```xml
                <info name="PCIVendor" value="HGST, Inc."/>
                <info name="PCIDevice" value="Ultrastar SN200 Series NVMe SSD"/>
```

Lassen:
```xml
              <info name="PCIVendor" value="Samsung Electronics Co Ltd"/>
              <info name="PCIDevice" value="NVMe SSD Controller 172Xa/172Xb"/>
```

Higher-level question, this is mainly an aesthetic thing, right?  Grabbing this information won't make a difference for scheduling?  If it doesn't matter for scheduling but we still want the nicely formed names, are you OK if we open an issue and handle this down the road?

SteVwonder replied:
Opened an issue to track this suggestion: https://github.com/flux-framework/flux-sched/issues/723

dongahn commented on t/Makefile.am:
```diff
@@ -64,6 +64,7 @@ TESTS = \
     t3023-resource-update2.t \
```
Super minor nit in commit message: "imperative" mood is preferred. Uses -> Use.

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
+
+test_description='Test node-local storage cases'
```
```suggestion
test_description='Test node-local storage cases with hwloc reader'
```

dongahn commented on t/data/resource/expected/node_local_storage/corona.R.out:
```diff
@@ -0,0 +1,37 @@
+      ------------core3[1:x]
+      ------------core4[1:x]
+      ------------core5[1:x]
+      ------------storage0[1562:x]
```
To confirm, does the entire block device (1562) has been allocated. Is the unit MB?

SteVwonder replied:
Yeah, the entire block device has been allocated.  The units are gBs (i.e., base 10); https://github.com/flux-framework/flux-sched/pull/714/files#diff-7dd3e7f6907b941d300d9ac623708499R263

The base 10 piece is because that is what storage device manufacturers use.  The Giga is because that seems like the right order of magnitude for user requests for storage (mB seems too small, and tB seems too big).

dongahn commented on t/data/resource/jobspecs/node_local_storage/corona.yaml:
```diff
@@ -0,0 +1,26 @@
+              - type: core
+                count: 24
+              - type: storage
+                count: 1500
```
It would be good to some more jobspec
1. requesting an entire device
2. requesting a very small amount (1 MB) -- this will make our granularity points clear in our test codes
3. if "chunk" is introduced, a case where multiple jobs that div up the storage

dongahn commented on t/data/resource/commands/node_local_storage/lassen.in:
```diff
@@ -0,0 +1,3 @@
+match allocate @TEST_SRCDIR@/data/resource/jobspecs/node_local_storage/lassen.yaml
```
Same comment on the commit msg. Uses -> Use;

dongahn commented on t/data/resource/commands/node_local_storage/catalyst.in:
```diff
@@ -0,0 +1,3 @@
+match allocate @TEST_SRCDIR@/data/resource/jobspecs/node_local_storage/catalyst.yaml
```
Commit msg.

dongahn commented on t/data/resource/expected/node_local_storage/catalyst.R.out:
```diff
@@ -0,0 +1,28 @@
+      ------------storage8[195:x]
+      ------------storage9[195:x]
+      ------------storage10[195:x]
+      ------------storage11[195:x]
```
Does Catalyst have multiple node-local block devices each of which was reported by hwloc as distinct storage?

SteVwonder replied:
Yeah. That's right.  See my comment above about how hwloc bubbles up what Linux considers to be block devices in `/sys/class/block/`.

dongahn commented on t/data/resource/expected/node_local_storage/catalyst.R.out:
```diff
@@ -10,10 +10,10 @@
       ------------core21[1:x]
       ------------core22[1:x]
       ------------core23[1:x]
+      ------------storage6[195:x]
```
This comment is not on this line, but this is the closest commit to post this comment (sorry).

The main question here is whether we want to rely on the logical indices that hwloc report as our storage id or use our own ID space. For example, at `walk_hwloc` we can add an id to each discovered device in the monotonically increasing manner to make the indices consistent across run to run, firmware update to update and hwloc version to version.

One down side though if flux exec is grown in the future where it can "contain" the block device size (using cgroup for example) as well, this scheme may or may not work.

Tagging @grondo. If you ever contain a job to node-local storage device resource, what mechanism you will be likely to use and what "handle" or "id" of a device do you need for that? 

If only the name of the device and the size are needed my proposed scheme should work.    

If you somehow will need hwloc to do this with its local object index, then we may be at the whim of hwloc OSDev indexing instability.

SteVwonder replied:
> If you ever contain a job to node-local storage device resource, what mechanism you will be likely to use and what "handle" or "id" of a device do you need for that?

If we use cgroups to contain node-local storage, it requires the `major:minor` of a device. We can access this via hwloc by accessing the attribute `LinuxDeviceID`.  Is there a "scratchpad" or extra space within each resource vertex that we can put that info (just for storage resources)?

Note that with cgroup containment, I think there are only two things you can contain:
1) Visibility of the device: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v1/devices.html
2) Bandwidth reading/writing to the device: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v1/blkio-controller.html

I don't see a way to divy the block device up capacity-wise with cgroups.  I think we would need partitioning on the device itself for that.

SteVwonder replied:
> The main question here is whether we want to rely on the logical indices that hwloc report as our storage id or use our own ID space. For example, at walk_hwloc we can add an id to each discovered device in the monotonically increasing manner to make the indices consistent across run to run, firmware update to update and hwloc version to version.

We can certainly override the numbering so that the counting starts at 0 and only increments by 1 for each storage device (rather than any OS device incrementing the hwloc logical ID).  That will handle some of the testing issues experienced in this PR.  In will also ensure that regardless if a job is allocated GPUs or not, the storage IDs will be the same.

I don't think that will buy us complete consistency across firmware updates or hwloc versions.  That method is beholden to the traversal order of hwloc.  For example, if a node has an NVMe drive and a SATA drive (like Catalyst does), in hwloc v1 the NVMe drives might be visited first and be labeled storage[0-3] and the SATA drive be visited next and labeled storage[4].  In hwloc v2, the SATA drives might be visited first and the NVMe drives next, resulting in storage[0] and storage[1-4]. So if we want to solve that, I think we will need to wait until a full traversal of the hwloc topology is complete, sort the drives based on something like serial number, then give them an ID.

SteVwonder replied:
Created an issue to track this suggestion/discussion: https://github.com/flux-framework/flux-sched/issues/724

dongahn commented:
Thanks @SteVwonder for this PR. This looks generally good. We need some more discussions for two areas:

1. Granularity of storage scheduling
2. Block device ID space. 

Please keep me honest about your L2 milestone due dates and effort/time requirements while discussing them, though. I will be open to landing some code with remaining workdocumented in issue tickets if you need some work to be merged for your L2 piece.

dongahn commented:
BTW, the topical commit structure and good commit messages make this review extremely easy and enjoyable!

SteVwonder commented:
> We need some more discussions for two areas:
>
>    Granularity of storage scheduling
>    Block device ID space.
>
> Please keep me honest about your L2 milestone due dates and effort/time requirements while discussing them, though. I will be open to landing some code with remaining workdocumented in issue tickets if you need some work to be merged for your L2 piece.

All good points. Thanks for the feedback @dongahn!  I left inline comments about both the granularity and the device IDs.  I don't think the former is achievable now without affecting the L2 deliverable.  Are you OK if we open an issue on that (and the associated memory granularity problem too)?  Some of the ID stuff is probably doable now. Let's iterate on that in the inline comments.

SteVwonder commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -247,8 +249,21 @@ void resource_reader_hwloc_t::walk_hwloc (resource_graph_t &g,
+            if (size_str == NULL || num_bytes == 0) {
+                break;
+            }
+            size = num_bytes / 1000000; // kBs -> gBs
```
Yeah, the size is always kB: https://www.open-mpi.org/projects/hwloc/doc/v2.0.1/a00319.php
Hwloc returns it as a string, but it is too large to fit into a 32bit int, as you suggest.  So I parsed it with strol.  I just realized I should have used `atol` rather than `strtol` (but not `atoi`).

SteVwonder replied:
Granted that this is only showing the user-accesible space, but the numbers line up pretty well for corona and catalyst.

Corona:
```
herbein1@corona110 ~
❯ df -h                                                                                                                                                                                                                         18:44:49 ()
Filesystem                                          Size  Used Avail Use% Mounted on
devtmpfs                                            126G     0  126G   0% /dev
tmpfs                                               252G     0  252G   0% /dev/shm
tmpfs                                               126G   19M  126G   1% /run
tmpfs                                               126G     0  126G   0% /sys/fs/cgroup
overlay                                             227G   98M  227G   1% /
tmpfs                                               227G   98M  227G   1% /overlay
/dev/nvme0n1                                        1.5T   33M  1.5T   1% /l/ssd
```
and `1562813784 / 1000000 = 1562.8gB = 1.5tB`

Catalyst:

```
❯ df -h                                                                                                                                                                                                                         18:47:47 ()
Filesystem                                          Size  Used Avail Use% Mounted on
devtmpfs                                             63G     0   63G   0% /dev
tmpfs                                               126G     0  126G   0% /dev/shm
tmpfs                                                63G   19M   63G   1% /run
tmpfs                                                63G     0   63G   0% /sys/fs/cgroup
overlay                                             113G   77M  113G   1% /
tmpfs                                               113G   77M  113G   1% /overlay
/dev/md0                                            745G   34M  745G   1% /l/ssd
```

Where `195360984 * 4 / 1000000 = 781.4gB`, and the `* 4` comes from the fact that there are 4 separate block devices.  And I think this one doesn't line up perfectly because of a small amount of overhead due to stitching 4 drives into a single logical mount.

SteVwonder replied:
> It appears to be GB given this line

Yeah, agreed.  hwloc reports kB, then I converted it to gB since that seems to be the nicest unit for user requests (kB is too small and tB is too big).  Let me know if you were expecting a different size.

> and the specs of the HGST NVMe model.

Based on [this line](https://github.com/flux-framework/flux-sched/blob/06ef8bc82c148c47ffd5716a0494c72e6e74b2cc/t/data/hwloc-data/001N/node_local_storage/corona/0.xml#L142), I believe corona has the SN200 model.  Based on the [SN200 data sheet](https://documents.westerndigital.com/content/dam/doc-library/en_us/assets/public/western-digital/product/data-center-drives/ultrastar-dc-ha200-series/data-sheet-ultrastar-dc-sn200.pdf), they do ship in 1.6TB packages (and they don't ship in 16 TB packages, that is a different model).  So I think we have the 1.6TB version and there is just some device overhead or formatting overhead that brings it down to 1.56TB.

SteVwonder replied:
>  I should have used atol rather than strtol (but not atoi).

I just realized why I didn't use atol:

> If the converted value would be out of the range of representable values by a long int, it causes undefined behavior. See strtol for a more robust cross-platform alternative when this is a possibility.

SteVwonder commented:
Thanks @dongahn and @milroy for the reviews and feedback.  I believe this PR is ready for another review. I addressed almost all of the comments OR opened an issue to track the continued discussion.  The only thing that I haven't resolved is the storing of the PCI device ID in the resource vertex (https://github.com/flux-framework/flux-sched/pull/714#discussion_r466723331).  If either of you can comment on the best place to store that information, I can probably get that done for this PR (assuming that metadata is also easily retrievable via the `dump` interface).

SteVwonder commented:
> If either of you can comment on the best place to store that information, I can probably get that done for this PR (assuming that metadata is also easily retrievable via the dump interface).

Thanks @dongahn for the suggestion to use the `properties` of each vertex.  I will quickly add the functionality to store the PCI device ID and then test that it is included in the emitted JGF.

SteVwonder commented:
Just pushed two commits that add support for capturing, storing and transmitting the `LinuxDeviceID` used by the kernel/cgroups for containment and throttling.  If Travis is happy, then I think this is ready for round 2 of reviews.

dongahn commented:
@SteVwonder: sounds all good. Do you want to fix two CI failures first? I will circle back to this once the CI is happy and give another review.

SteVwonder commented:
> Do you want to fix two CI failures first?

Had a broken shell `&&` chain in one of the new tests.  Pushed a fixup commit, and most of Travis is happy now.  Getting an error in a presumably unrelated recovery test on Centos.  Going to kick that builder in the hopes that it is an intermittent failure (since it passed previously).

```
[0;31mFAIL[m: t1007-recovery-full.t 6 - recovery: works when both modules restart (rv1)
[0;31mFAIL[m: t1007-recovery-full.t 7 - recovery: a cancel leads to a job schedule (rv1)
[0;31mFAIL[m: t1007-recovery-full.t 8 - recovery: works when only qmanager restarts (rv1)
[0;31mFAIL[m: t1007-recovery-full.t 9 - recovery: a cancel leads to a job schedule (rv1)
[0;31mFAIL[m: t1007-recovery-full.t 10 - recovery: both modules restart (rv1->rv1_nosched)
[0;31mFAIL[m: t1007-recovery-full.t 11 - recovery: only qmanager restarts (rv1->rv1_nosched)
[0;31mFAIL[m: t1007-recovery-full.t 12 - recovery: cancel all jobs (rv1_nosched)
[0;31mFAIL[m: t1007-recovery-full.t 13 - recovery: restart w/ no running jobs (rv1_nosched)
[0;31mFAIL[m: t1007-recovery-full.t 14 - recovery: submit to occupy resources fully (rv1_nosched)
[0;31mFAIL[m: t1007-recovery-full.t 15 - recovery: qmanager restarts (rv1_nosched->rv1_nosched)
[0;31mFAIL[m: t1007-recovery-full.t 16 - recovery: a cancel leads to a job schedule (rv1_nosched)
[0;31mFAIL[m: t1007-recovery-full.t 17 - recovery: cancel all jobs (rv1_nosched)
[0;31mFAIL[m: t1007-recovery-full.t 18 - removing resource and qmanager modules
[0;31mFAIL[m: t1007-recovery-full.t 19 - load sched-simple module
     t1007-recovery-full.t: [0;31mFAIL[m: N=19  PASS=5   FAIL=14 SKIP=0 XPASS=0 XFAIL=0
<snip>
ok 5 - recovery: cancel one running job without fluxion
recovery: cancel one running job without fluxion

expecting success: 
    reload_resource load-allowlist=node,core,gpu match-format=rv1 &&
    reload_qmanager &&
    test_must_fail flux ion-resource info ${jobid1} &&
    flux ion-resource info ${jobid2} | grep "ALLOCATED" &&
    flux ion-resource info ${jobid3} | grep "ALLOCATED" &&
    flux ion-resource info ${jobid4} | grep "ALLOCATED" &&
    flux job wait-event -t 10 ${jobid5} start &&
    test_expect_code 3 flux ion-resource info ${jobid6}

FileNotFoundError: error(2): Unable to connect to Flux: ENOENT: No such file or directory
not ok 6 - recovery: works when both modules restart (rv1)

expecting success: 
    flux job cancel ${jobid2} &&
    flux job wait-event -t 10 ${jobid6} start

flux-job: flux_open: No such file or directory
not ok 7 - recovery: a cancel leads to a job schedule (rv1)
<snip>
2020-08-11T18:02:40.329018Z job-exec.info[0]: 19914555392: 0: stderr: flux-shell: ERROR: affinity: affinity: core4 not in topology
2020-08-11T18:02:40.365629Z job-manager.debug[0]: submit_cb: added 1 jobs
2020-08-11T18:02:40.368862Z sched-fluxion-qmanager.info[0]: f=0x7fa6ac0277f0
2020-08-11T18:02:40.374356Z connector-local.debug[0]: subscribe shell-21072183296.
2020-08-11T18:02:40.392431Z job-exec.info[0]: 21072183296: 0: stderr: flux-shell: ERROR: affinity: affinity: core1 not in topology
2020-08-11T18:02:40.418250Z job-manager.debug[0]: submit_cb: added 1 jobs
2020-08-11T18:02:40.419145Z sched-fluxion-qmanager.info[0]: f=0x7fa6ac0086a0
2020-08-11T18:02:40.454732Z job-manager.debug[0]: submit_cb: added 1 jobs
2020-08-11T18:02:40.455636Z sched-fluxion-qmanager.info[0]: f=0x7fa6ac03d610
2020-08-11T18:02:40.511912Z broker.debug[0]: rmmod sched-fluxion-qmanager
2020-08-11T18:02:40.512357Z sched-fluxion-qmanager.debug[0]: service_unregister
2020-08-11T18:02:40.512305Z job-manager.debug[0]: alloc: stop due to alloc response error: Function not implemented
2020-08-11T18:02:40.512674Z broker.debug[0]: module sched-fluxion-qmanager exited
2020-08-11T18:02:40.512905Z sched-fluxion-resource.debug[0]: disconnect_request_cb: a notify request aborted
2020-08-11T18:02:40.526467Z broker.debug[0]: rmmod sched-fluxion-resource
2020-08-11T18:02:40.526944Z broker.debug[0]: module sched-fluxion-resource exited
2020-08-11T18:02:40.527171Z resource.debug[0]: acquire_disconnect: resource.acquire aborted
2020-08-11T18:02:40.541213Z job-exec.debug[0]: exec aborted: id=18035507200
2020-08-11T18:02:40.541304Z job-exec.debug[0]: exec_kill: 18035507200: signal 15
2020-08-11T18:02:40.548010Z connector-local.debug[0]: unsubscribe shell-18035507200.
2020-08-11T18:02:40.755308Z broker.debug[0]: insmod sched-fluxion-resource
2020-08-11T18:02:40.756091Z sched-fluxion-resource.debug[0]: mod_main: resource module starting
2020-08-11T18:02:40.761514Z sched-fluxion-resource.info[0]: populate_resource_db: loaded resources from hwloc in the KVS
2020-08-11T18:02:40.761859Z sched-fluxion-resource.debug[0]: resource status changed (rankset=[all] status=DOWN)
2020-08-11T18:02:40.761915Z sched-fluxion-resource.debug[0]: resource status changed (rankset=[0] status=UP)
2020-08-11T18:02:40.761953Z sched-fluxion-resource.debug[0]: mod_main: resource graph database loaded
2020-08-11T18:02:40.825557Z broker.debug[0]: insmod sched-fluxion-qmanager
2020-08-11T18:02:40.826430Z sched-fluxion-qmanager.debug[0]: enforced policy (queue=default): fcfs
2020-08-11T18:02:40.826496Z sched-fluxion-qmanager.debug[0]: effective queue params (queue=default): default
2020-08-11T18:02:40.826538Z sched-fluxion-qmanager.debug[0]: effective policy params (queue=default): default
2020-08-11T18:02:40.827653Z sched-fluxion-qmanager.debug[0]: handshaking with sched-fluxion-resource completed
2020-08-11T18:02:40.827802Z sched-fluxion-qmanager.debug[0]: service_register
2020-08-11T18:02:40.827926Z job-manager.debug[0]: scheduler: hello
2020-08-11T18:02:40.829766Z sched-fluxion-qmanager.debug[0]: requeue success (queue=default id=18035507200)
2020-08-11T18:02:40.831426Z sched-fluxion-qmanager.debug[0]: requeue success (queue=default id=18907922432)
```

SteVwonder commented:
It was an intermittent failure.  Travis is happy now.

dongahn commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -274,6 +278,14 @@ int resource_reader_hwloc_t::walk_hwloc (resource_graph_t &g,
+                // Non-fatal error, just log a warning
+                m_err_msg += "Error getting the LinuxDeviceID of block storage device; ";
+            } else {
+                properties["LinuxDeviceID"] = devID_str;
```
It would be a bit safer with `insert`
```suggestion
                auto ret = properties.insert (
                               std::pair<std::string,
                                         std::string> ("LinuxDeviceID",
                                                       devID_str));
                if (!ret) {
                    errno = EEXIST;
                    m_err_msg += "Error inserting LinuxDeviceID into properties map; ";
                    rc = -1;
                    break;
                }

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -9,6 +9,13 @@ exp_dir="${SHARNESS_TEST_SRCDIR}/data/resource/expected/node_local_storage"
 xml_dir="${SHARNESS_TEST_SRCDIR}/data/hwloc-data/001N/node_local_storage/"
 query="../../resource/utilities/resource-query"
 
+extract_storage_ids() {
```
Probably need:

`skip_all_unless_have jq` in the beginning of this test since `jq` is required for these tests.

Or use `HAVE_JQ` on the tests that use `extract_storage_ids` function?

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -29,6 +36,14 @@ test_expect_success "${corona_small_desc}" '
 
+corona_LinuxDeviceID_desc="ensure that PCI device ID included in resource vertex and JGF output"
+test_expect_success "${corona_LinuxDeviceID_desc}" '
+    ${query} -d -L ${corona_xml} -f hwloc -W node,socket,core,gpu,storage -S CA -P low \
```
Long line.

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -29,6 +36,14 @@ test_expect_success "${corona_small_desc}" '
+test_expect_success "${corona_LinuxDeviceID_desc}" '
+    ${query} -d -L ${corona_xml} -f hwloc -W node,socket,core,gpu,storage -S CA -P low \
+-t corona-LinuxDeviceID.jgf.out -F jgf < corona_full_cmds &&
+    extract_storage_ids corona-LinuxDeviceID.jgf.out corona-LinuxDeviceID.out
```
Probably need to wrap

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -29,6 +36,14 @@ test_expect_success "${corona_small_desc}" '
+    ${query} -d -L ${corona_xml} -f hwloc -W node,socket,core,gpu,storage -S CA -P low \
+-t corona-LinuxDeviceID.jgf.out -F jgf < corona_full_cmds &&
+    extract_storage_ids corona-LinuxDeviceID.jgf.out corona-LinuxDeviceID.out
+    test_cmp ${exp_dir}/corona-LinuxDeviceID.out corona-LinuxDeviceID.out
```
Same

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -48,6 +63,14 @@ test_expect_success "${lassen_small_desc}" '
     test_cmp ${exp_dir}/lassen-small-drive.R.out lassen-small-drive.R.out
 '
 
+lassen_LinuxDeviceID_desc="ensure that PCI device ID included in resource vertex and JGF output"
```
The test name is same as above and the line exceeds 80?

`lassen_LinuxDeviceID_desc="ensure that PCI device ID is properly set 2"`

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -48,6 +63,14 @@ test_expect_success "${lassen_small_desc}" '
+    ${query} -d -L ${lassen_xml} -f hwloc -W node,socket,core,gpu,storage -S CA -P low \
+-t lassen-LinuxDeviceID.jgf.out -F jgf < lassen_full_cmds &&
+    extract_storage_ids lassen-LinuxDeviceID.jgf.out lassen-LinuxDeviceID.out
+    test_cmp ${exp_dir}/lassen-LinuxDeviceID.out lassen-LinuxDeviceID.out
```
Some long lines.

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -67,4 +90,12 @@ test_expect_success "${catalyst_small_desc}" '
     test_cmp ${exp_dir}/catalyst-small-drive.R.out catalyst-small-drive.R.out
 '
 
+catalyst_LinuxDeviceID_desc="ensure that PCI device ID included in resource vertex and JGF output"
```
The test name is same as above and the line exceeds 80?

`catalyst_LinuxDeviceID_desc="ensure that PCI device ID is properly set 3"`

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -67,4 +90,12 @@ test_expect_success "${catalyst_small_desc}" '
+    ${query} -d -L ${catalyst_xml} -f hwloc -W node,socket,core,gpu,storage -S CA -P low \
+-t catalyst-LinuxDeviceID.jgf.out -F jgf < catalyst_full_cmds &&
+    extract_storage_ids catalyst-LinuxDeviceID.jgf.out catalyst-LinuxDeviceID.out
+    test_cmp ${exp_dir}/catalyst-LinuxDeviceID.out catalyst-LinuxDeviceID.out
```
Some lines are long

dongahn commented on t/t3026-resource-node-local-storage.t:
```diff
@@ -37,10 +37,10 @@ test_expect_success "${corona_small_desc}" '
 
 corona_LinuxDeviceID_desc="ensure that PCI device ID included in resource vertex and JGF output"
-test_expect_success "${corona_LinuxDeviceID_desc}" '
+test_expect_success HAVE_JQ "${corona_LinuxDeviceID_desc}" '
```
Ah... Nice to add `HAVE_JQ`. Please disregard my previous comments then.

SteVwonder commented on resource/readers/resource_reader_hwloc.cpp:
```diff
@@ -274,6 +278,14 @@ int resource_reader_hwloc_t::walk_hwloc (resource_graph_t &g,
             supported_resource = true;
+
+            const char* devID_str = hwloc_obj_get_info_by_name (obj, "LinuxDeviceID");
+            if (devID_str == NULL) {
+                // Non-fatal error, just log a warning
+                m_err_msg += "Error getting the LinuxDeviceID of block storage device; ";
```
> or just get rid of this warning message and instead add a "fake" devID_str to the property map?

Sounds good to me.  I just have it setting the ID to `-1` now.

SteVwonder commented:
@dongahn, thanks for the feedback!  I fixed the long lines in the test, switched to using the `insert` method for the `properties` map, and replaced the logging with a `-1` LinuxDeviceID.  Let me know if you have any additional feedback.  If not, I'll squash the fixups.  Thanks!
